sql-plugin/src/main/scala/com/nvidia/spark/rapids/InputFileBlockRule.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable.{ArrayBuffer, LinkedHashMap}
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, InputFileBlockLength, InputFileBlockStart, InputFileName}
+import org.apache.spark.sql.execution.{FileSourceScanExec, LeafExecNode, SparkPlan}
+import org.apache.spark.sql.execution.datasources.v2.BatchScanExec
+import org.apache.spark.sql.execution.exchange.ShuffleExchangeExec
+
+/**
+ * InputFileBlockRule is to prevent the SparkPlans
+ * [SparkPlan (with first input_file_xxx expression), FileScan) to run on GPU
+ *
+ * See https://github.com/NVIDIA/spark-rapids/issues/3333
+ */
+object InputFileBlockRule {
+
+  private def checkHasInputFileExpressions(exec: Expression): Boolean = exec match {
+    case _: InputFileName => true
```

Comment:
With this we now have two different ways of checking for this situation and there is nothing forcing the two together.  At a minimum can we add in comments here and on `GpuExpression.disableCoalesceUntilInput` pointing to each other so we know to try and keep the two in sync? It would be even better if we could come up with a way through tags or something that would make it very difficult for someone to add in a new method that needs this and not update both places.  Although I don't think more methods are coming, so the comments are good enough.

Reply:
I suppose you were meaning `GpuTransitionOverrides.disableCoalesceUntilInput`. I just rename and put the same function in the GpuTransitionOverrides.

Reply:
`GpuExpression.disableCoalesceUntilInput` is called from `GpuExec.disableCoalesceUntilInput` which in turn is called from `GpuTransitionOverrides.insertCoalesce` to figure out if it should be disabled or not for instances of `GpuExec`. For things that are not instances of `GpuExec` the code calls into `GpuTransitionOverrides.disableCoalesceUntilInput`, which was updated to call into `GpuTransitionOverrides.checkHasInputExpressions`.  It would be nice to have a way to keep them in sync so we only have to update things in one place. Now that I look more closely there is no good way to do a comment for this.  I think we could clean up the code more to remove duplication between the two paths, but I think this is fine for now.